### PR TITLE
Minor: Improve docstrings for `LogicalPlan`

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

n/a

## Rationale for this change
While working on https://github.com/influxdata/influxdb_iox/pull/8515 I had some questions about the various `LogicalPlan` that I had to look up in the code. I wanted to encode this information as doc comments so I didn't have to do it a second time

## What changes are included in this PR?
1. Update docstrings for `LogicalPlan`

## Are these changes tested?
N/A

## Are there any user-facing changes?
better (hopefully) docs

### CODE REVIEW

File: datafusion/expr/src/logical_plan/plan.rs [hunk 1]
```diff
@@ -69,63 +69,84 @@ pub enum LogicalPlan {
...
+    /// `INTERSECT [ALL]`.
     Union(Union),
-    /// Produces rows from a table provider by reference or from the context
+    /// Produces rows from a [`TableSource`], used to implement SQL
+    /// `FROM` tables or views.
     TableScan(TableScan),
-    /// Produces no rows: An empty relation with an empty schema
+    /// Produces no rows: An empty relation with an empty schema that
+    /// produces 0 or 1 rows. This is used to implement SQL `SELECT`
+    /// that has no values in the `FROM` clause.
```
Comment (metesynnada) [inline datafusion/expr/src/logical_plan/plan.rs:107]:
```suggestion
    /// Produces no rows: An empty relation with an empty schema that
    /// produces 0 or 1 row. This is used to implement SQL `SELECT`
    /// that has no values in the `FROM` clause.
```

File: datafusion/expr/src/logical_plan/plan.rs [hunk 2]
```diff
@@ -69,63 +69,84 @@ pub enum LogicalPlan {
...
     Values(Values),
     /// Produces a relation with string representations of
-    /// various parts of the plan
+    /// various parts of the plan. This is used to implement SQL `EXPLAIN`.
     Explain(Explain),
-    /// Runs the actual plan, and then prints the physical plan with
-    /// with execution metrics.
+    /// Runs the input, and prints annotated physical plan as a string
+    /// with with execution metric. This is used to implement SQL
+    /// `EXPLAIN ANALYZE`.
```
Comment (metesynnada) [inline datafusion/expr/src/logical_plan/plan.rs:128]:
```suggestion
    /// Runs the input, and prints annotated physical plan as a string
    /// with execution metric. This is used to implement SQL
    /// `EXPLAIN ANALYZE`.
```

File: datafusion/expr/src/logical_plan/plan.rs [hunk 3]
```diff
@@ -69,63 +69,84 @@ pub enum LogicalPlan {
...
+    /// Extension operator defined outside of DataFusion. This is used
+    /// to extend DataFusion with custom relational operations that
     Extension(Extension),
-    /// Remove duplicate rows from the input
+    /// Remove duplicate rows from the input. This is used to
+    /// implement SQL `SELECT DISTINCT ...`.
     Distinct(Distinct),
-    /// Prepare a statement
+    /// Prepare a statement and find any bind parameters
+    /// (e.g. `?`). This is used to implement SQL prepared statements.
```
Comment (metesynnada) [inline datafusion/expr/src/logical_plan/plan.rs:137]:
```suggestion
    /// Prepare a statement and find any bind parameters
    /// (e.g. `?`). This is used to implement SQL-prepared statements.
```

File: datafusion/expr/src/logical_plan/plan.rs [hunk 4]
```diff
@@ -69,63 +69,84 @@ pub enum LogicalPlan {
...
     Dml(DmlStatement),
-    /// CREATE / DROP TABLES / VIEWS / SCHEMAs
+    /// Data Definition Language (DDL): CREATE / DROP TABLES / VIEWS / SCHEMAs
     Ddl(DdlStatement),
-    /// COPY TO
+    /// `COPY TO` for writing plan results to files
     Copy(CopyTo),
-    /// Describe the schema of table
+    /// Describe the schema of table. This is used to implement the
+    /// SQL `DESCRIBE` command from MySQL.
```
Comment (metesynnada) [inline datafusion/expr/src/logical_plan/plan.rs:146]:
```suggestion
    /// Describe the schema of the table. This is used to implement the
    /// SQL `DESCRIBE` command from MySQL.
```

**Review by metesynnada — COMMENTED** [review]
There are a few minor typos. Thank you for the helpful explanations.

**Review by alamb — COMMENTED** [review]
Thank you @metesynnada 🙏

**Review by yjshen — APPROVED** [review]

**Review by jackwener — APPROVED** [review]

**Comment by alamb** [comment]
Thanks @jackwener and @jackwener

**Pull request merged**